Adding odf 4.12 deployment automation

aditidukle opened the pull request:
@sudeeshjohn For automation to work we need to add the following credentials to jenkins:
1. ODF_IBMCLOUD_API_KEY
2. PULL_SECRET --> Add odf related tokens to the existing pull secret
3. ODF_AUTH_YAML

aditidukle commented:
/assign @sudeeshjohn

Shilpi-Das1 commented on jobs/pipelines/powervs/odf/4.12/daily-odf4.12-powervs-lon06-tier-2/Jenkinsfile:
```diff
@@ -0,0 +1,175 @@
+	 }
+    
+    stages {
+        stage('Clone ansible extra') {
```
@aditidukle If you are not using this stage you can remove it

Shilpi-Das1 commented on vars/odfCaptureTierSummary.groovy:
```diff
@@ -0,0 +1,21 @@
+def call(){
```
@aditidukle Since these are all using shell commands you can create shell script under dir /scripts/odf  and directly call them from Jenkins file

Shilpi-Das1 replied:
@aditidukle This comment is not addressed

Shilpi-Das1 commented on vars/odfRunDeployScript.groovy:
```diff
@@ -0,0 +1,19 @@
+def call(){
```
@aditidukle Since these are all using shell commands you can create shell script under dir /scripts/odf and directly call them from Jenkins file

Shilpi-Das1 replied:
@aditidukle This comment is not addressed

Shilpi-Das1 commented on vars/odfRunSetupScript.groovy:
```diff
@@ -0,0 +1,34 @@
+def call(){
```
@aditidukle Since these are all using shell commands you can create shell script under dir /scripts/odf and directly call them from Jenkins file No need to add methods in vars file

Shilpi-Das1 replied:
@aditidukle this comment is not addressed

Shilpi-Das1 commented on vars/odfRunTierTests.groovy:
```diff
@@ -0,0 +1,20 @@
+def call(){
```
Same script comment

Shilpi-Das1 replied:
@aditidukle This comment is not addressed

Shilpi-Das1 commented on vars/setupCommonEnvironmentVariables.groovy:
```diff
@@ -4,15 +4,15 @@ def call() {
         //VMs setup
         if ( env.POWERVS == "true" ) {
-            env.INSTANCE_NAME = "rdr-cicd"
+            //env.INSTANCE_NAME = "rdr-cicd"
```
Remove comment

Shilpi-Das1 commented on vars/setupCommonEnvironmentVariables.groovy:
```diff
@@ -4,15 +4,15 @@ def call() {
             env.ENABLE_LOCAL_REGISTRY = "false"
             env.LOCAL_REGISTRY_IMAGE = "docker.io/ibmcom/registry-ppc64le:2.6.2.5"
-            env.SETUP_SQUID_PROXY = "true"
+            //env.SETUP_SQUID_PROXY = "true"
```
Remove comment

Shilpi-Das1 commented on jobs/pipelines/powervs/odf/4.12/daily-odf4.12-powervs-syd05-tier-4c/Jenkinsfile:
```diff
@@ -0,0 +1,175 @@
+	 }
+    
+    stages {
+        stage('Clone ansible extra') {
```
Remove this stage if not needed

Shilpi-Das1 commented on jobs/pipelines/powervs/odf/4.12/daily-odf4.12-powervs-syd05-tier-3/Jenkinsfile:
```diff
@@ -0,0 +1,175 @@
+	 }
+    
+    stages {
+        stage('Clone ansible extra') {
```
Remove this stage if not needed

Shilpi-Das1 commented on jobs/pipelines/powervs/odf/4.12/daily-odf4.12-powervs-sao01-tier-4a/Jenkinsfile:
```diff
@@ -0,0 +1,174 @@
+                deployCluster()
+            }
+        }
+        stage('Run crontab script for capturing outputs of multiple commands') {
```
Remove this stage as well if not needed. Please check in all Jenkins file

Shilpi-Das1 reviewed (changes requested):
Also Please add schedule in hack/jjb_template.jinja2

sudeeshjohn commented on deploy/templates/.deploy-openshift4-powervs-script-odf.tfvars.template:
```diff
@@ -0,0 +1,50 @@
+bastion                     = {memory      = "__BASTION_MEMORY__",   processors  = "__BASTION_PROCESSORS__", "count"   = 1}
+bootstrap                   = {memory      = "__BOOTSTRAP_MEMORY__",   processors  = "__BOOTSTRAP_PROCESSORS__",  "count"   = 1}
+master                      = {memory      = "__MASTER_MEMORY__",   processors  = "__MASTER_PROCESSORS__",  "count"   = __NUM_OF_MASTERS__}
+worker                      = {memory      = "__WORKER_MEMORY__",   processors  = "__WORKER_PROCESSORS__",  "count"   = __NUM_OF_WORKERS__,  data_volume_count = 1,  data_volume_size = 500 }
```
why can't we make the data volume size configurable ? and the volume count too?

sudeeshjohn commented on deploy/templates/.deploy-openshift4-powervs-script-odf.tfvars.template:
```diff
@@ -0,0 +1,50 @@
+setup_squid_proxy = "__SETUP_SQUID_PROXY__"
+
+storage_type                = "notnfs"
+setup_snat                  = true
```
Why can't we make use of `deploy-openshift4-powervs.tfvars.template` it self ?

sudeeshjohn commented on deploy/makefiles/Makefile.openshift4_pvs_script_odf:
```diff
@@ -0,0 +1,32 @@
+deploy\:openshift4\:powervs\:script\:odf\:clean: %clean: %destroy
+ifeq ($(shell test -d $(OPENSHIFT_POWERVS_SCRIPT_DEPLOY_DIR) && echo -n yes),yes)
+	@rm -rf $(OPENSHIFT_POWERVS_SCRIPT_DEPLOY_DIR)
+endif
```
can we make use of `Makefile.openshift4_pvs_script` itself ?

sudeeshjohn commented on vars/setupCommonEnvironmentVariables.groovy:
```diff
@@ -48,21 +48,26 @@ def call() {
-            if (OCP_RELEASE == "4.5" || OCP_RELEASE == "4.6" ) {
-                env.E2E_EXCLUDE_LIST = "https://raw.github.ibm.com/redstack-power/e2e-exclude-list/${env.OCP_RELEASE}-powervm/ocp${env.OCP_RELEASE}_power_exclude_list.txt"
+            if ( env.ODF_VERSION ) {
+                env.INSTANCE_NAME = "jenkins-odf"
```
Stick to rdr-cicd at the starting,this to identify the CI resources..  something like `rdr-cicdodf` ?

sudeeshjohn commented on deploy/Makefile:
```diff
@@ -167,6 +167,44 @@ deploy-openshift4-powervs-script\:output:
+	@$(SELF) -s deploy:openshift4:powervs:script:odf:clean OPENSHIFT_POWERVS_SCRIPT_DEPLOY_DIR=$(TERRAFORM_DIR) OPENSHIFT_POWERVS_SCRIPT_TERRAFORM_VARS_FILE=$(TERRAFORM_VARS_FILE)
 
 
 # ---------------------------------
```
This additional section in the Makefile can be avoided if makes of same template and makefile

sudeeshjohn commented on deploy/Makefile:
```diff
@@ -167,6 +167,44 @@ deploy-openshift4-powervs-script\:output:
+.PHONY: deploy-openshift4-powervs-script-odf\:clean
+## Clean up power-openstack build artifacts
+deploy-openshift4-powervs-script-odf\:clean:
+	@$(SELF) -s deploy:openshift4:powervs:script:odf:clean OPENSHIFT_POWERVS_SCRIPT_DEPLOY_DIR=$(TERRAFORM_DIR) OPENSHIFT_POWERVS_SCRIPT_TERRAFORM_VARS_FILE=$(TERRAFORM_VARS_FILE)
```
This additional section in the Makefile can be avoided if makes of same template and makefile

aditidukle commented:
@Shilpi-Das1 @sudeeshjohn All review comments have been incorporated. Could you please check again?

Shilpi-Das1 commented:
@aditidukle Some of the comments are not addressed Also can you please do a single commit

aditidukle commented:
@Shilpi-Das1 I have made the following changes for your comments on functions:
odfCaptureTierSummary.groovy --> Since this function only had 2 shell commands, I added it in the jenkinsfile itself instead of creating a seperate shell script  in scripts/odf
odfRunDeployScript.groovy --> same as above
odfRunTierTests.groovy --> same as above
odfRunSetupScript.groovy --> I created a script scripts/odf/odf-setup-script.sh and called it from the jenkinsfile

sudeeshjohn commented on vars/setupClusterConfig.groovy:
```diff
@@ -28,6 +30,23 @@ def call(String config, String noOfWorkers="2"){
+        else if (env.ODF_VERSION) {
+            env.BASTION_MEMORY = "16"
+            env.BASTION_PROCESSORS = "1"
+
```
do we really need these much resources for bastion ?

sudeeshjohn replied:
@aditidukle do we need 1 proc for bastion ?

sudeeshjohn replied:
you can go with .5 for bastion

aditidukle commented:
Tested PR changes on molly server - http://molly.aus.stglabs.ibm.com:8090/job/powervs/job/odf/job/test-pr-job/28/ 
Made following changes after testing:
- Removed tier test 4a as all tests from tier 4a are not distributed to tier 2, 4b, and 4c
- Added a new config called "odf" in setupClusterConfig.groovy

Shilpi-Das1 reviewed (approved):
/lgtm

sudeeshjohn reviewed (approved):
/lgtm
